text/0000-read-buf.md

```diff
@@ -0,0 +1,624 @@
+        }
+    }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+```rust
+/// A wrapper around a byte buffer that is incrementally written to and initialized.
+///
+/// This type is a sort of "double cursor". It tracks three regions in the buffer: a region at the beginning of the
+/// buffer that has been logically written to, a region that has been initialized at some point but not yet logically
+/// written to, and a region at the end that is fully uninitialized. The written-to region is guaranteed to be a
+/// subset of the initialized region.
+pub struct ReadBuf<'a> {
+    buf: &'a mut [MaybeUninit<u8>],
+    written: usize,
+    initialized: usize,
+}
+
+impl<'a> ReadBuf<'a> {
+    /// Creates a new `ReadBuf` from a fully initialized buffer.
+    #[inline]
+    pub fn new(buf: &'a mut [u8]) -> ReadBuf<'a> { ... }
+
+    /// Creates a new `ReadBuf` from a fully uninitialized buffer.
+    ///
+    /// Use `assume_initialized` if part of the buffer is known to be already inintialized.
+    #[inline]
+    pub fn uninit(buf: &'a mut [MaybeUninit<u8>]) -> ReadBuf<'a> { ... }
+
+    /// Returns the full size of the buffer.
+    #[inline]
+    pub fn len(&self) -> usize { ... }
+
+    /// Returns a shared reference to the written-to portion of the buffer.
+    #[inline]
+    pub fn written(&self) -> &[u8] { ... }
+
+    /// Returns a mutable reference to the written-to portion of the buffer.
+    #[inline]
+    pub fn written_mut(&mut self) -> &mut [u8] { ... }
+
+    /// Returns a shared reference to the initialized portion of the buffer.
+    ///
+    /// This includes the written-to portion.
+    #[inline]
+    pub fn initialized(&self) -> &[u8] { ... }
+
+    /// Returns a mutable reference to the initialized portion of the buffer.
+    ///
+    /// This includes the written-to portion.
+    #[inline]
+    pub fn initialized_mut(&mut self) -> &mut [u8] { ... }
+
+    /// Returns a mutable reference to the unwritten-to part of the buffer without ensuring that it has been fully
+    /// initialized.
+    ///
+    /// # Safety
+    ///
+    /// The caller must not de-initialize portions of the buffer that have already been initialized.
+    #[inline]
+    pub unsafe fn unwritten_mut(&mut self) -> &mut [MaybeUninit<u8>] { ... }
+
+    /// Returns a mutable reference to the unwritten-to part of the buffer, ensuring it is fully initialized.
+    ///
+    /// Since `ReadBuf` tracks the region of the buffer that has been initialized, this is effectively "free" after
+    /// the first use.
+    #[inline]
+    pub fn initialize_unwritten(&mut self) -> &mut [u8] { ... }
+
+    /// Returns a mutable reference to the first `n` bytes of the unwritten-to part of the buffer, ensuring it is
+    /// fully initialized.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `self.remaining()` is less than `n`.
+    #[inline]
+    pub fn initialize_unwritten_to(&mut self, n: usize) -> &mut [u8] { ... }
```

Comment:
To me, the "to" here sounds like it defines the target of the operation, like "write X to Y". Also the relation with "up to index `n`" is a bit weak since this is not index `n` in the buffer, it is index `n` *in the uninitialized part* of the buffer (I think).

Maybe `initialize_unwritten_n` or `initialize_unwritten_len` or `initialize_unwritten_with_len` or so?

Reply:
Maybe a better set of name for the three methods that give access to the unwritten part would be `unwritten`, unwritten_len` (both of which return `&mut [u8]`), and `unwritten_maybe_uninit` (returning `&mut [MaybeUninit<u8>]`).

Reply:
I do think it's somewhat important for the names of these methods to indicate that they're potentially doing significant amounts of work.

Reply:
Hm, that's fair. OTOH, having the most dangerous method have the shortest name (`unwritten_mut`) also doesn't seem great.